core/che-core-tracing-core/src/main/java/org/eclipse/che/core/tracing/TracingInterceptor.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.core.tracing;
+
+import com.google.common.annotations.Beta;
+import com.google.inject.Inject;
+import io.opentracing.Scope;
+import io.opentracing.Tracer;
+import java.lang.reflect.Method;
+import java.util.Map;
+import java.util.WeakHashMap;
+import java.util.function.Supplier;
+import org.aopalliance.intercept.MethodInterceptor;
+import org.aopalliance.intercept.MethodInvocation;
+import org.eclipse.che.commons.annotation.Traced;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Beta
+public class TracingInterceptor implements MethodInterceptor {
+
+  private static final Logger LOG = LoggerFactory.getLogger(TracingInterceptor.class);
+
+  private Tracer tracer;
+  private WeakHashMap<Class<?>, WeakHashMap<Method, String>> spanNames = new WeakHashMap<>();
+
+  @Inject
+  public void init(Tracer tracer) {
+    this.tracer = tracer;
+  }
+
+  @Override
+  public Object invoke(MethodInvocation invocation) throws Throwable {
+    String spanName = getSpanName(invocation);
+    try (Scope scope =
+        tracer.buildSpan(spanName).asChildOf(tracer.activeSpan()).startActive(true)) {
+
+      Traced.TagsStack.push();
+
+      try {
+        return invocation.proceed();
+      } finally {
+        for (Map.Entry<String, Supplier<?>> e : Traced.TagsStack.pop().entrySet()) {
+          Object val;
+          try {
+            val = e.getValue().get();
+          } catch (Exception ex) {
+            if (LOG.isDebugEnabled()) {
+              // we want to know the exception in case the tag extraction failed.
+              // Slf4j doesn't seem to provide a method overload that could both provide formatting
+              // arguments and the cause.
+              LOG.debug(
+                  "Could not get the value for a tag called "
+                      + e.getKey()
+                      + " when tracing "
+                      + spanName
+                      + ".",
```

Comment:
You could use `String.format()` here.

Reply:
Or even `LOG.debug("blah-blah {}: {}", 1, 2, ex);`

Reply:
> Or even LOG.debug("blah-blah {}: {}", 1, 2, ex);

Does that work? I wasn't sure it does...